### PR TITLE
Skybox color customization options

### PR DESCRIPTION
**Pull request by logarrhythmic**
Current state of these options:
Adds a custom sky color picker, and a dropdown to replace the sky color in either the overworld or everywhere. I should probably change this to a toggle and custom color picker for overworld and caves separately, to accommodate both #7299 and #9159. 

Old text:

> Addresses #9159, although further options might still be nice and the current UX feels a bit strange.
> 
> Adds the following options to the skybox plugin:
> 
> *Custom sky color* - standard color picker
> 
> *Mode* - dropdown, "None", "Replace target sky color", "Overworld" and "Everywhere". *None* is the current behavior of using the sky map and making no substitutions. *Replace target sky color* uses the custom color in place of a specific color value, and blending between areas still works properly. *Overworld* uses the custom color when the player is not in an underground area, and *Everywhere* always uses the custom color.
> 
> *Set current sky as target* - checkbox, gives an "are you sure" window and then stores the sky color of the player's current chunk as the target color. Automatically resets to unchecked, but currently the config panel doesn't update to reflect this until it is reopened. Let me know if there's a way to force an update, and I'll fix this PR to do that. Alternatively, if there's a way to add a button to the config panel that triggers the onConfigChanged event or something similar, that would be exactly what I'm looking for here.
> 
> Target color - hidden option, used as storage and set by the above checkbox.
> 
> The idea with the target color feature is to allow for replacement of the most common sky color, which I personally find too gray, while still getting the special sky colors of various swamps, volcanoes, caves, and other areas which have received special attention in the skybox definition file.
> 
> I also considered a feature that would allow for several source -> replacement pairs to be set, but I couldn't think of a good way to make the UI for that. If there's interest, I'll consider revisiting that idea.

### CODE REVIEW

**Comment by abextm** [comment]
I really don't like how this abuses the config

**Comment by logarrhythmic** [comment]
same

**Comment by logarrhythmic** [comment]
I see there's a framework for adding entire panels, like the library plugin does. Would it be okay to add one of those? It seems like a pretty good solution to me, and it would also let me add functionality for multiple replacement colors instead of just the 1.
I'm thinking I would replace the "target color" mode with an "advanced" option in the dropdown, which would enable the sky color replacement panel.
The panel would then have a more sophisticated UI with a list of color pairs, and a way to add new ones.

**Comment by abextm** [comment]
There isn't a framework for adding custom things to the config panel, which is what this really needs. It would probably be best for this to just be a global override color that makes it not invoke `Skybox` entirely

**Comment by logarrhythmic** [comment]
I could strip this PR down to just have the global overrides and then work on the replacement pairs to make another PR later, since I really want that specific feature myself.

**Comment by abextm** [comment]
that would probably be best. I think the best thing for more granular overrides would be to be able to merge the config in our `skybox.txt` format

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java [hunk 1]
```diff
@@ -91,6 +103,17 @@ public void onBeforeRender(BeforeRender r)
 			return;
 		}
 
+		if
+		(
+			config.overrideMode() == SkyOverrideMode.ALL ||
+			(config.overrideMode() == SkyOverrideMode.OVERWORLD && client.getLocalPlayer().getWorldLocation().getY() < 4200)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java]: It looks to me like `4160` marks the first tile beyond the northernmost water, and this should also be defined in some publicly-accessible location (maybe `Constants`?) for general use--in the past, other plugins/features have wanted a reference to this value.

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/Skybox.java [hunk 2]
```diff
@@ -352,14 +352,17 @@ private int chunkData(int cx, int cy, int plane, ChunkMapper chunkMapper)
 		}
 
 		int cv = chunks[(stride * (cy - y1)) + (cx - x1)];
+
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/skybox/Skybox.java]: Now that there are no more code changes to Skybox.java, remove the added newlines in this file.

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java]: Remove the newline between the copyright header and package definition.

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+package net.runelite.client.plugins.skybox;
+
+import net.runelite.client.config.Config;
+import net.runelite.client.config.ConfigGroup;
+import net.runelite.client.config.ConfigItem;
+
+import java.awt.Color;
+import net.runelite.client.plugins.skybox.config.SkyOverrideMode;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java]: All imports should be in a single block with no newlines separating them.

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java [hunk 5]
```diff
@@ -91,6 +103,17 @@ public void onBeforeRender(BeforeRender r)
...
+		)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java]: The formatting for this condition doesn't quite match what we use elsewhere in the project. I'd expect this to look like the following:

```java
if (config.overrideMode() == SkyOverrideMode.ALL
	|| (config.overrideMode() == SkyOverrideMode.OVERWORLD && client.getLocalPlayer().getWorldLocation.getY() < 4200))
{
	// ...
}
```

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java (hunk 1 above)
Comment (logarrhythmic) [reply runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java:109]: Yup, seems good, adding it as Constants.OVERWORLD_MAX_Y.

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/Skybox.java (hunk 2 above)
Comment (logarrhythmic) [reply runelite-client/src/main/java/net/runelite/client/plugins/skybox/Skybox.java:356]: Yep, will do

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java (hunk 3 above)
Comment (logarrhythmic) [reply runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java:25]: Alright

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java (hunk 4 above)
Comment (logarrhythmic) [reply runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPluginConfig.java:33]: Okay

File: runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java (hunk 5 above)
Comment (logarrhythmic) [reply runelite-client/src/main/java/net/runelite/client/plugins/skybox/SkyboxPlugin.java:110]: Okay

**Comment by logarrhythmic** [comment]
I'm going to change this to two toggles and custom color pickers for overworld and caves separately, to accommodate both #7299 and #9159.

**Pull request merged**